reset.c

```diff
@@ -0,0 +1,888 @@
+        {
+            if (resetconf.keepPGs <= 0) {
+                usage(stderr);              //already given!
+                return EXIT_FAILURE;
+            }
+            if (optarg) {                   //filename given, reject selected PGs only
+                pgrejname = optarg;
+                resetconf.keepPGs = 0;
+            }
+            else {
+                resetconf.keepPGs = -1;     //reject all PGs -- reset execution will anyway be added
+            }
+        }
+        break;
+        case 'i':                           //input file name
+        {
+            if (inname) {                   //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+            }
+            inname = optarg;
+        }
+        break;
+        case 'o':                           //output file name
+        {
+            if (outname) {                  //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+
+            }
+            outname = optarg;
+        }
+        break;
+        case 'k':                           //keep aux tags file name
+        {
+            if (auxkeepname) {              //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+
+            }
+            auxkeepname = optarg;
+        }
+        break;
+        case 'r':                           //rejected aux tags file name
+        {
+            if (auxrejname) {               //already given!
+                usage(stderr);
+                return EXIT_FAILURE;
+            }
+            auxrejname = optarg;
+        }
+        break;
+        // handle standard samtool options like thread count, verbosity...
+        default:
+            if (parse_sam_global_opt(c, optarg, lopts, &ga) == 0)
+                break;
+            // else fall-through
+        // couldn't parse or unknown options, show usage!
+        case '?':                           //unknown options found!
+        {
+            usage(stderr);
+            return EXIT_FAILURE;
+            break;
+        }
+
+        }
+    }
+
+    //check and fail if unnecessary parameters are given
+    if ((argc - optind) > 0 || !inname || !outname)
+    {
+        usage(stderr);
+        return EXIT_FAILURE;
+    }
+
+    //get the aux tags to be retained
+    if (getauxtags(&resetconf, auxkeepname, auxrejname))
+        return EXIT_FAILURE;
+
+    //get the pg names to be discarded
+    if (getrejectedpgnames(&resetconf, pgrejname))
+        return EXIT_FAILURE;
+
+    //set output file format based on name
+    sam_open_mode(outmode + 1, outname, NULL);
+
+    //open input and output files
+    infile = sam_open(inname, "r");
+    outfile = sam_open_format(outname, outmode, &ga.out);
+    if (!infile || !outfile) {
+        printf("Could not open %s%s%s", !infile ? inname : "", (!infile && !outfile)? ", " : "", !outfile ? outname : "");
```

Comment:
There are lots of `printf` calls for error messages.  These need to be on stderr and not stdout (see `-o file` vs stdout mentioned elsewhere.

However you may wish to look into `print_error` and `print_error_errno` calls instead.  The latter is best used in examples such as the `sam_open_format` where `errno` will have been set, so an appropriate system error message may be produced (eg "no such file or directory" or "permission denied").